### PR TITLE
build(declarations): create `schema-to-declarations` script add declarations

### PR DESCRIPTION
**Pull request by TheLarkInn**
**What kind of change does this PR introduce?**
Build related change

**Did you add tests for your changes?**
N/A (?)

**If relevant, link to documentation update:**
N/A (?)

**Summary**
Since half of the webpack codebase is referencing `this.options` or the webpack options in some form of another. This means that to do this we would need to hand write a declaration file for the entire webpack options based on our `json` schema.

Introducing an amazing module called `json-schema-to-typescript` which allow us to simply pipe in the schema file. 

TODO: I don't know how I want to incorporate the creation of these declarations to be automatic (maybe like lint-staged and husky?)

**Does this PR introduce a breaking change?**
Nope!

**Other information**

### CODE REVIEW

**Comment by mohsen1** [comment]
Why you decided to go from JSON Schema to TS instead of other way around?

**Comment by TheLarkInn** [comment]
JSON Schema can still be the source of truth (as it always is for the config validation). The value here is that we just instantly TypeDef'd the entire `this.options` object via generation of code we already have.

File: declarations/WebpackOptions.d.ts [hunk 1]
```diff
@@ -0,0 +1,1249 @@
...
+		| (
+				| "web"
+				| "webworker"
+				| "node"
+				| "async-node"
+				| "node-webkit"
+				| "electron-main"
+				| "electron-renderer")
+		| {
+				[k: string]: any;
```
Comment (TheLarkInn) [inline declarations/WebpackOptions.d.ts]: This is invalid and should be corresponding to: 

Original property:

Comment (TheLarkInn) [reply declarations/WebpackOptions.d.ts]: cc @bcherny, this looks like this might be a bug, I'll submit and reference this comment.

Comment (TheLarkInn) [reply declarations/WebpackOptions.d.ts]: Tied now to bcherny/json-schema-to-typescript#164

File: .github/ISSUE_TEMPLATE.md [hunk 2]
```diff
@@ -1,22 +0,0 @@
-<!-- Please don't delete this template or we'll close your issue -->
```
Comment (ooflorent) [inline .github/ISSUE_TEMPLATE.md]: Why did you remove these files?

File: declarations/WebpackOptions.d.ts [hunk 3]
```diff
@@ -0,0 +1,1249 @@
...
+	wrappedContextCritical?: boolean;
+	/**
+	 * Enable recursive directory lookup for partial dynamic dependencies
+	 */
+	wrappedContextRecursive?: boolean;
+	/**
+	 * Set the inner regular expression for partial dynamic dependencies
+	 */
+	wrappedContextRegExp?: {
+		[k: string]: any;
```
Comment (ooflorent) [inline declarations/WebpackOptions.d.ts]: This is wrong

File: package.json [hunk 4]
```diff
@@ -1,170 +1,181 @@
...
-      "\\.runtime\\.js$",
-      "<rootDir>/test/",
-      "<rootDir>/schemas/"
-    ],
-    "testEnvironment": "node",
-    "coverageReporters": [
-      "json"
-    ]
-  }
+	"name": "webpack",
```
Comment (ooflorent) [inline package.json]: Could you revert the formatting changes here?

File: declarations/WebpackOptions.d.ts (hunk 3 above)
Comment (ooflorent) [reply declarations/WebpackOptions.d.ts]: All `RegExp` are missing

**Comment by mohsen1** [comment]
Since this is more than just json schema I suggest using a runtime type checker that builds on top of TypeScript types and write the schemas in TypeScript instead.  

That said runtime type checker that uses TypeScript types does not exist as far as I know.

File: .github/ISSUE_TEMPLATE.md (hunk 2 above)
Comment (sokra) [reply .github/ISSUE_TEMPLATE.md]: He is still learning git^^

File: declarations/WebpackOptions.d.ts (hunk 3 above)
Comment (TheLarkInn) [reply declarations/WebpackOptions.d.ts]: Yup this is something we will have to support in the transformation. I've spoke with the maintainer of the json-schema-to-typescript package and there is a way we could support these extra ajv types.

File: .github/ISSUE_TEMPLATE.md (hunk 2 above)
Comment (TheLarkInn) [reply .github/ISSUE_TEMPLATE.md]: LOL @sokra ever since these got changed, I've had line ending warnings and these get thrown into every PR. I'm not sure why this keeps happening so yes I'm still learning git :joy:

**Comment by jeremejevs** [comment]
The more automation, the better!

However, even if `json-schema-to-typescript` adds support for extensions, how will you go about typing functions? Right now it's just an opaque `"instanceof": "Function"` everywhere, TS can do much better than that. Functions are already typed in `DefinitelyTyped`, to a varying degree of fidelity, so switching to something like `(...args: any[]) => any` would be a downgrade. Need some sort of annotation.

Ideally, I'd go with what @mohsen1 suggests, and generate the schema from annotated TypeScript instead. I'm using [`ts-json-schema-generator`](https://github.com/vega/ts-json-schema-generator) for something very similar in my own project. But I'm not 100% sure that you can express any conceivable schema that way, especially with extensions, and it's going to be quite an undertaking, considering that Webpack is pure JS. Is introduction of TS even on the table?

**Comment by TheLarkInn** [comment]
Perhaps we do the reverse? When I can jump back into this I'll look into it!

File: declarations/WebpackOptions.d.ts (hunk 1 above)
Comment (sokra) [reply declarations/WebpackOptions.d.ts]: I created https://github.com/bcherny/json-schema-to-typescript/pull/168 to address this.

**Comment by sokra** [comment]
Blocked by https://github.com/bcherny/json-schema-to-typescript/pull/191

File: package.json [hunk 5]
```diff
@@ -17,6 +17,7 @@
     "enhanced-resolve": "^4.1.0",
     "eslint-scope": "^4.0.0",
     "json-parse-better-errors": "^1.0.2",
+    "json-schema-to-typescript": "^6.0.1",
```
Comment (kuba-kubula) [inline package.json]: Hi @sokra and @TheLarkInn,

shouldn't this package be added only into `devDependencies`?
Because of `json-schema-to-typescript` package being present now in dependencies, its own dependencies are also installed, examples:

```
 "@types/cli-color": "^0.3.29",
    "@types/json-schema": "^7.0.0",
    "@types/lodash": "^4.14.116",
    "@types/minimist": "^1.2.0",
    "@types/mz": "0.0.32",
    "@types/node": "^10.9.4",
    "@types/prettier": "^1.13.2",
```

I don't think it was the intent here, or am I missing something?

Comment (pksunkara) [reply package.json]: I agree, it's only used in tooling. So, it should be in `devDependencies`.

Comment (sokra) [reply package.json]: Yep you are right, fixing it.

Comment (sokra) [reply package.json]: https://github.com/webpack/webpack/pull/8068

File: lib/WebpackOptionsApply.js [hunk 6]
```diff
@@ -450,7 +462,11 @@ class WebpackOptionsApply extends OptionsApply {
 		}
 		if (options.optimization.minimize) {
 			for (const minimizer of options.optimization.minimizer) {
-				minimizer.apply(compiler);
+				if (typeof minimizer === "function") {
+					minimizer.apply(compiler);
+				} else {
+					minimizer.apply(compiler);
+				}
```
Comment (jdelStrother) [inline lib/WebpackOptionsApply.js]: @sokra is this a typo or am I missing something?  There's a similar bit further down with `plugin.apply(compiler)`
